core/trino-main/src/main/java/io/trino/operator/output/RowPositionsAppender.java

```diff
@@ -0,0 +1,236 @@
+
+import io.trino.spi.block.Block;
+import io.trino.spi.block.RowBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import io.trino.spi.type.RowType;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.trino.operator.output.PositionsAppenderUtil.calculateBlockResetSize;
+import static io.trino.operator.output.PositionsAppenderUtil.calculateNewArraySize;
+import static io.trino.spi.block.RowBlock.createRowBlockInternal;
+import static java.util.Objects.requireNonNull;
+
+public class RowPositionsAppender
+        implements PositionsAppender
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(RowPositionsAppender.class).instanceSize();
+    private final PositionsAppender[] fields;
+    private int initialEntryCount;
+    private boolean initialized;
+
+    private int positionCount;
+    private boolean hasNullRow;
+    private boolean hasNonNullRow;
+    private boolean[] rowIsNull = new boolean[0];
+    private int[] fieldBlockOffsets = new int[1];
+    private long retainedSizeInBytes;
+    private long sizeInBytes;
+
+    public static RowPositionsAppender createRowAppender(
+            PositionsAppenderFactory positionsAppenderFactory,
+            RowType type,
+            int expectedPositions,
+            long maxPageSizeInBytes)
+    {
+        PositionsAppender[] fields = new PositionsAppender[type.getFields().size()];
+        for (int i = 0; i < fields.length; i++) {
+            fields[i] = positionsAppenderFactory.create(type.getFields().get(i).getType(), expectedPositions, maxPageSizeInBytes);
+        }
+        return new RowPositionsAppender(fields, expectedPositions);
+    }
+
+    private RowPositionsAppender(PositionsAppender[] fields, int expectedPositions)
+    {
+        this.fields = requireNonNull(fields, "fields is null");
+        this.initialEntryCount = expectedPositions;
+        updateRetainedSize();
+    }
+
+    @Override
+    public void append(IntArrayList positions, Block block)
+    {
+        if (positions.isEmpty()) {
+            return;
+        }
+        ensureCapacity(positions.size());
+        RowBlock sourceRowBlock = (RowBlock) block;
+        IntArrayList notNullPositions;
+        if (sourceRowBlock.mayHaveNull()) {
+            notNullPositions = processNullablePositions(positions, sourceRowBlock);
+            hasNullRow |= notNullPositions.size() < positions.size();
+            hasNonNullRow |= notNullPositions.size() > 0;
+        }
+        else {
+            // the source Block does not have nulls
+            notNullPositions = positions;
+
+            for (int i = 0; i < positions.size(); i++) {
+                fieldBlockOffsets[positionCount + i + 1] = fieldBlockOffsets[positionCount] + i + 1;
+            }
+            hasNonNullRow = true;
+        }
+
+        sourceRowBlock.writeRowsTo(
```

Comment:
Have we considered using `ColumnarRow#toColumnarRow` instead of adding new methods to SPI ?

Reply:
Not sure if I understand this correctly but I think that would be problematic. Imagine if `append` gets only the last position in the `RowBlock`. To calculate its field position I would have to iterate over all positions in the `RowBlock` to see which one is null. This would basically recalculate `io.trino.spi.block.RowBlock#fieldBlockOffsets` for every `append`

Reply:
Okay, I see the problem with that.
Another idea : How about we add a method `Block getRowBlockField(int index, int[] positions)` to ColumnarRow. This can copy fieldBlockOffsets for the required positions and produce a wrapper Block which delegates to original RowBlock for isNull and for getXXX APIs uses the fieldBlockOffsets to jump to right position in the primitive fields block.